### PR TITLE
DOC: Improve replace docstring

### LINKED ISSUE #13852: Series.replace and DataFrame.replace have same docstring?
```
In [4]: pd.__version__
Out[4]: u'0.18.1'
```

The docstring for Series.replace refers to looking up column names, which AFAIK doesn't make sense for the Series version of `replace`.  It seems like the Series function is just blindly inheriting the DataFrame docstring?

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.replace.html?highlight=replace#pandas.Series.replace

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.replace.html?highlight=replace#pandas.DataFrame.replace

For example, the following trivial example of `Series.replace()` does not agree with the docstring:

```
import pandas as pd

x = pd.DataFrame([dict(A=1, B=2, C=3), dict(A=2, B=3, C=4), dict(A=3, B=4, C=5)])

print(x.A.replace({1:0}))
print(x.A)

0    0
1    2
2    3
Name: A, dtype: int64
0    1
1    2
2    3
Name: A, dtype: int64
```

### LINKED ISSUE #17673: DOC: 'replace' docstring lacking / too complex
I think the `replace` docstring is lacking in many ways (https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.replace.html):

- The explanation of `to_replace` keyword is both way too complex and lacking an explanation of the simple cases:
  - the most simple case of a scalar value is not clearly mentioned (like `df.replace(to_replace=0, replace=1)`, it is mentioned in the 'str' explanation, but it is not specific to strings)
  - the simplest dict case of `df.replace({to_replace: replacement})` is not mentioned (the dict explanation starts with explanation of nested dicts)
  - I would personally rewrite this whole explanation of this keyword, start with basic cases, and only after that (or in the notes) explain the complex cases.
- There is a reference to the examples section for "examples of each of those", but there is no examples section. We should add one.
- In the 'see also' section it references `reindex`, `asfreq` and `fillna`. `fillna` is fine, but I fail to see the link with the first two. I would rather add a reference to `where` to replace values based on a boolean condition (and the 'see also' should not just refer to the other methods, but also include a sentence on why / the difference)
- The docstring also uses `NDFrame`, and this should never be in a public docstring (failing substituion of docstring in `generic`)
- I would personally also write separate docstrings for the series and dataframe case. This will give some duplication, but I think this gives room to simplify the docstring (or certainly for the simpler Series.replace case). (xref https://github.com/pandas-dev/pandas/issues/13852)

See the tutorial docs (https://pandas.pydata.org/pandas-docs/stable/missing_data.html#replacing-generic-values) with some actual examples.

Underlying reason is that this function of course can do way too many things at the same time (or the same things in too many different ways) ... (orthogonal to this, we could maybe also think if certain functionality could be moved into its own function).

### PR DESCRIPTION
**Pull request by reidy-p**
xref #17673
closes #13852
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

I have tried to separate the docstrings for series and dataframe as suggested in the original issue but can try to combine them if that works better.

As noted in the original issue, ``replace`` can do a lot things and I've tried to cover most of them with the examples but would welcome other suggestions.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -227,6 +232,30 @@
 
 """
 
+def _single_replace(self, to_replace, method, inplace, limit):
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Could define this in one place and then import it in `frame.py` and `series.py`.

Also a short docstring would still be nice, even though it's internal.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3030,6 +3059,364 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
+            string. Alternatively, this could be a regular expression or a list,
+            dict, or array of regular expressions in which case ``to_replace``
+            must be ``None``.
+        method : string, optional, {'pad', 'ffill', 'bfill'}
+            The method to use when for replacement, when ``to_replace`` is a
+            ``list``.
+
+        See Also
+        --------
+        :func:`DataFrame.fillna` : Fill NA/NaN values
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: You do't need the

   :func:`

. Just DataFrame.fillna should be enough for numpydoc.

**Review by TomAugspurger — COMMENTED** [review]
Haven't gone through this in detail yet, but I think it'd be cleaner to leave the implementation in `pandas/core/generic.py`, and to only overwrite the docstring in `series.py` and `frame.py`.

So in, e.g. `series.py` it'd look like

```python
    def replace(self, to_replace=None, value=None, inplace=False, limit=None,
                regex=False, method='pad', axis=None):
   """add your docstring"""
    super().replace(to_replace= to_replace, value=value...)
```

Then you don't have all the duplicated code.

**Comment by reidy-p** [comment]
@TomAugspurger thanks for the review, made some updates.

**Comment by jreback** [comment]
can you rebase

File: pandas/core/series.py [hunk 3]
```diff
@@ -2658,6 +2661,193 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
                                           limit=limit, downcast=downcast,
                                           **kwargs)
 
+    def replace(self, to_replace=None, value=None, inplace=False, limit=None,
```
Comment (jreback) [inline pandas/core/series.py]: can we share *most* / all of this via ``_shared_docs``? otherwise end up with lots of duplication here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by reidy-p** [comment]
I removed the duplication by using ``_shared_docs`` but as a result most of the examples for both the ``Series.replace`` and ``DataFrame.replace`` docstrings now use ``DataFrame``.

**Comment by jreback** [comment]
@jorisvandenbossche if you'd have a look. I only glanced briefly :>

**Review by jreback — APPROVED** [review]

File: pandas/core/generic.py [hunk 4]
```diff
@@ -4787,94 +4791,111 @@ def bfill(self, axis=None, inplace=False, limit=None, downcast=None):
...
 
-                - First, if `to_replace` and `value` are both lists, they
+                - First, if ``to_replace`` and ``value`` are both lists, they
                   **must** be the same length.
                 - Second, if ``regex=True`` then all of the strings in **both**
                   lists will be interpreted as regexs otherwise they will match
-                  directly. This doesn't matter much for `value` since there
+                  directly. This doesn't matter much for ``value`` since there
                   are only a few possible substitution regexes you can use.
                 - str and regex rules apply as above.
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: I suppose this is "str, regex, and numeric rules apply as above." now?

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**